File: core/src/main/java/com/pluxity/station/LineService.java
package com.pluxity.station;

import static com.pluxity.global.constant.ErrorCode.DUPLICATE_LINE_NAME;
import static com.pluxity.global.constant.ErrorCode.NOT_FOUND_LINE;

import com.pluxity.global.exception.CustomException;
import com.pluxity.global.utils.SortUtils;
import com.pluxity.station.dto.LineCreateRequest;
import com.pluxity.station.dto.LineResponse;
import com.pluxity.station.dto.LineUpdateRequest;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class LineService {

    private final LineRepository lineRepository;
    private StationService stationService;

    @Autowired
    public void setStationService(@Lazy StationService stationService) {
        this.stationService = stationService;
    }

    public LineService(LineRepository lineRepository) {
        this.lineRepository = lineRepository;
    }

    @Transactional
    public Long save(LineCreateRequest request) {
        lineRepository
                .findByName(request.name())
                .ifPresent(
                        line -> {
                            throw new CustomException(DUPLICATE_LINE_NAME, request.name());
                        });
        Line line = Line.builder().name(request.name()).color(request.color()).build();
        return lineRepository.save(line).getId();
    }

    @Transactional(readOnly = true)
    public List<LineResponse> findAll() {
        return lineRepository.findAll(SortUtils.getOrderByCreatedAtDesc()).stream()
                .map(LineResponse::from)
                .toList();
    }

    @Transactional(readOnly = true)
    public LineResponse findById(Long id) {
        Line line = lineRepository.findById(id).orElseThrow(() -> notFoundException(id));
        return LineResponse.from(line);
    }

    @Transactional(readOnly = true)
    public Line findLineById(Long id) {
        return lineRepository.findById(id).orElseThrow(() -> notFoundException(id));
    }

    @Transactional(readOnly = true)
    public List<Long> findStationsByLineId(Long lineId) {
        Line line = findLineById(lineId);
        return line.getStations().stream().map(Station::getId).collect(Collectors.toList());
    }

    @Transactional
    public void update(Long id, LineUpdateRequest request) {
        Line line = findLineById(id);
        line.update(Line.builder().name(request.name()).color(request.color()).build());
    }

    @Transactional
    public void delete(Long id) {
        Line line = findLineById(id);

        List<Station> stations = new ArrayList<>(line.getStations());
        for (Station station : stations) {
            station.removeLine(line);
        }

        lineRepository.delete(line);
    }

    @Transactional
    public void addStationToLine(Long lineId, Long stationId) {
        Line line = findLineById(lineId);
        Station station = stationService.findStationById(stationId);
        station.addLine(line);
    }

    private static CustomException notFoundException(Long id) {
        return new CustomException(NOT_FOUND_LINE, id);
    }
}

File: core/src/main/java/com/pluxity/station/LineService.kt
package com.pluxity.station

import com.pluxity.global.constant.ErrorCode
import com.pluxity.global.exception.CustomException
import com.pluxity.global.utils.SortUtils
import com.pluxity.station.dto.LineCreateRequest
import com.pluxity.station.dto.LineResponse
import com.pluxity.station.dto.LineUpdateRequest
import com.pluxity.station.dto.toLineResponse
import lombok.RequiredArgsConstructor
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@RequiredArgsConstructor
class LineService(
    private val lineRepository: LineRepository,
    private val stationLineService: StationLineService,
) {
    @Transactional
    fun save(request: LineCreateRequest): Long {
        lineRepository
            .findByName(request.name)?.let {
                throw CustomException(ErrorCode.DUPLICATE_LINE_NAME, request.name)
            }
        val line =
            Line(
                name = request.name,
                color = request.color,
            )
        return lineRepository.save(line).id!!
    }

    @Transactional(readOnly = true)
    fun findAll(): List<LineResponse> = lineRepository.findAll(SortUtils.getOrderByCreatedAtDesc()).map { it.toLineResponse() }

    @Transactional(readOnly = true)
    fun findById(id: Long): LineResponse = (lineRepository.findByIdOrNull(id) ?: throw notFoundException(id)).toLineResponse()

    @Transactional(readOnly = true)
    fun findLineById(id: Long): Line = lineRepository.findByIdOrNull(id) ?: throw notFoundException(id)

    @Transactional(readOnly = true)
    fun findStationsByLineId(lineId: Long): List<Long> = findLineById(lineId).getStations().map { it.id }

    @Transactional
    fun update(
        id: Long,
        request: LineUpdateRequest,
    ) {
        val line = findLineById(id)

        if (request.name != null) {
            lineRepository
                .findByNameAndIdNot(request.name, id)
                ?.let {
                    throw CustomException(ErrorCode.DUPLICATE_LINE_NAME, request.name)
                }
        }

        line.update(request.name, request.color)
    }

    @Transactional
    fun delete(id: Long) {
        val line = findLineById(id)

        val stations: List<Station> = line.getStations()
        for (station in stations) {
            stationLineService.deleteStationLine(station, line)
        }

        lineRepository.delete(line)
    }

    private fun notFoundException(id: Long): CustomException {
        return CustomException(ErrorCode.NOT_FOUND_LINE, id)
    }
}
